### PR TITLE
Make a PDDL ontology with shapes that is machine processable

### PR DESCRIPTION
**Pull request by likelion**
Apparently, this PR also contains a PDDL generator, a Planner OSLC Service (based on Metric-FF planner), and a number of useful OSLC features explained below.

### CODE REVIEW

**Comment by likelion** [comment]
I think, I reached a "logical comma" in refining our PDDL ontology. Few design points:

* This document was used as a reference for a PDDL dialect:
  https://www.jair.org/media/1129/live-1129-2132-jair.pdf
* There are no variables per se. Variables only appear because they describe parameters of predicates, functions or actions. Hence, there is only `Parameter` class.
* Let's define terminology (using pseudo programming language):
  * `function myfunc(A: int, B:string) {...}` here **A** and **B** are **parameters**.
  * `int Y; X := myfunc(Y, "5")` here **Y** and **"5"** are **arguments**.
* Parameters are _orderable_ typed _properties_, i.e. if we have a bunch of parameters - their mutual order is important, they have an associated type, and they appear as properties (**predicates** in subject-**predicate**-object relation) of `Predicate`, `Function` or `Action`.
* Order of parameters is modelled using SHACL `order` property at the moment. There are at least two other ways of modelling the order:
  * Using `rdf:List`s and combination of `rdf:first`, `rdf:rest` properties
  * Using `rdf:Seq` and corresponding `rdf:_1`, `rdf_2`, ... properties
* There may be multiple usages of the same user-defined _type_, _predicate_, _function_ or _action_. Hence, they are _meta classes_, i.e. they are first subclassed by a **user** (to create corresponding definitions) and then subclasses are instantiated multiple times for usage (in case of types, instantiations are objects). In fact, `Action` is subclassed by a **user** and instantiated not by a **user**, but by a **PDDL Planner**. All other classes in the ontology are only instantiated (not subclassed).
* There is no need to model the order of arguments of `Predicate`, `Function` or `Action` because arguments are always bound to their corresponding parameters (which themselves are already ordered) by appearing as **objects** in the subject-predicate(parameter)-**object**(argument) relations.
* Arguments of built-in operators such as `>`, `imply`, `/`, etc. are also order sensitive. For them there are two properties defined: `pddl:left` and `pddl:right` that allow to specify left and right arguments respectively. For the order insensitive operators (such as `and`) the arguments are specified using `pddl:argument` property.
* OSCL shapes, unfortunately, do not allow to completely validate PDDL documents, however partial validation is possible by using `oslc:occurs` and `oslc:range` checks (it is not always possible to specify precise `oslc:range` types as some of them are user-defined, and inferencing over `rdfs:subClassOf` is not allowed in OSLC).

Here is the current PDDL ontology (see pddl.ttl):
diagram-2

**Comment by likelion** [comment]
Here is usage of that ontology to specify an example planning domain:
diagram-2

and example planning problem:

diagram-3

File: oslc_prolog/lib/oslc.pl [hunk 1]
```diff
@@ -316,11 +317,16 @@
 %                      in which > and \ are \-escaped. */
 %    ==
 
+copy_resource([], [], _, _, _) :- !.
+
+copy_resource([H|T], [H|T], Source, Sink, Options) :- !,
```
Comment (likelion) [inline oslc_prolog/lib/oslc.pl:329]: FEATURE: copy lists of resources, not just a single resource.

File: oslc_prolog/lib/oslc_core.pl [hunk 2]
```diff
@@ -48,12 +47,46 @@
...
+    -> findall(Class, (
+         rdfs_individual_of(IRI, Class),
+         IRI \= Class
+       ), Classes),
+       append(L4, Classes, L5)
+    ; L5 = L4
+    ),
+    list_to_set(L5, Set)
+  ; Options = [],
+    Set = []
```
Comment (likelion) [inline oslc_prolog/lib/oslc_core.pl:102]: FEATURE: OSLC Core: query any resource via GET request with the following parameters:
* `<class>?rdfs=sup` - to get all super classes of the `<class>` also
* `<class>?rdfs=sub` - to get all subclasses of the `<class>` also
* `<class>?rdfs=ind` - to get all individuals of the `<class>` also
* `<individual>?rdfs=cls` - to get all classes the `<individual>` is inheriting also

File: oslc_prolog/lib/oslc_rdf.pl [hunk 3]
```diff
@@ -47,6 +47,25 @@
...
+  -> fail
+  ; oslc:unmarshal_property(IRI, PropertyDefinition, Value, Type, rdf(Graph))
+  ).
+
+oslc:delete_property(IRI, PropertyDefinition, tmp(Graph)) :-
+  ( var(Graph)
+  -> make_temp_graph(Graph)
+  ; oslc:delete_property(IRI, PropertyDefinition, rdf(Graph))
+  ).
+
```
Comment (likelion) [inline oslc_prolog/lib/oslc_rdf.pl:68]: FEATURE: use `tmp(Graph)` as source/sink to create temporary named graphs. If used in resource handlers, temporary graphs are volatile (non-persistent), if created by a serving thread are automatically garbage collected after the response is sent.

File: planner_reasoner/lib/pddl_generator.pl [hunk 4]
```diff
@@ -0,0 +1,479 @@
+:- module(pddl_generator, [
+    generate_pddl/2
+]).
+
+:- use_module(library(semweb/rdf11)).
+:- use_module(library(semweb/rdfs)).
+
+generate_pddl(Resource, Stream) :-
```
Comment (likelion) [inline planner_reasoner/lib/pddl_generator.pl]: FEATURE: generate PDDL text for either domain or problem `Resource` into the `Stream`

File: oslc_prolog/lib/oslc_core.pl [hunk 5]
```diff
@@ -19,24 +20,38 @@
...
+                 [dcterms:created='^^'(T, xsd:dateTime)], tmp(Context.graph_out)).
 
 get_time_class(Context) :-
-  make_temp_graph(Context.graph_out),
-  create_resource(oslc:'Time', [rdfs:'Class'], [], rdf(Context.graph_out)).
+  create_resource(oslc:'Time', [rdfs:'Class'], [], tmp(Context.graph_out)).
 
 handle_ontology(Context) :-
-  once((
+  findall(Graph, (
```
Comment (likelion) [inline oslc_prolog/lib/oslc_core.pl:30]: FEATURE: When GETting an ontology by `http://<host>/<prefix>`, and multiple named graphs have a name that starts with URI of the `<prefix>`, the response will contain resources from all of them (see `pddle` ontology for example).

File: oslc_prolog/lib/oslc_dispatch.pl [hunk 6]
```diff
@@ -11,10 +11,16 @@
...
   response/1,
-  response/2
+  response/2,
+  select_acceptable_content_type/2
 ]).
 
 :- use_module(library(semweb/rdf11)).
+:- use_module(library(semweb/turtle)).
+
+:- multifile serializer/2.
```
Comment (likelion) [inline oslc_prolog/lib/oslc_dispatch.pl:21]: FEATURE: Users can extend accepted MIME types by specifying custom `serializer/2` and `serialize_response/3` to provide custom serialization of response graph.

File: oslc_prolog/lib/oslc_dispatch.pl [hunk 7]
```diff
@@ -11,10 +11,16 @@
...
+:- multifile serialize_response/3.
+:- multifile error_message/2.
```
Comment (likelion) [inline oslc_prolog/lib/oslc_dispatch.pl:23]: FEATURE: Users can extend supported messages explaining different error HTTP status codes by specifying custom `error_message/2`

File: oslc_prolog/lib/oslc_dispatch.pl [hunk 8]
```diff
@@ -69,15 +75,22 @@
...
   ), Handlers),
-  predsort(handler_compare, Handlers, [handler(Method, IRI, Module:Predicate, _)|_]),
+  predsort(handler_compare, Handlers, SortedHandlers),
+  rdf_global_id(Prefix:Resource, IRI),
   NewContext = Context.put(iri, IRI).put(iri_spec, Prefix:Resource),
-  T =.. [Predicate, NewContext],
-  call(Module:T).
+  dispatch_to_handlers(NewContext, SortedHandlers).
+
+dispatch_to_handlers(Context, [handler(Module:Predicate, _)|T]) :-
```
Comment (likelion) [inline oslc_prolog/lib/oslc_dispatch.pl:87]: FEATURE: if a resource handler returns `false`, dispatcher will call the next handler with lower priority that matches the resource; if no handlers left - OSLC Core will handle it.

**Comment by likelion** [comment]
I believe this PR closes issues #31, #32, #33, and PR #30.

**Comment by swarupmohalik** [comment]
[ a oslc:Property ;
                  dcterms:description "Problem init."^^rdf:XMLLiteral ;
                  oslc:propertyDefinition :init ;
                  oslc:name "init"^^xsd:string ;
                  oslc:occurs oslc:Zero-or-many ;
                  oslc:valueType oslc:AnyResource

**There should be exactly one occurrence of :init**

:FluentMinusShape
  a oslc:ResourceShape ;
  oslc:describes :Minus ;
  dcterms:description "Unary operator (-) in fluent."^^rdf:XMLLiteral ;
  oslc:property [ a oslc:Property ;
                  dcterms:description "Unary operator argument (fluent)."^^rdf:XMLLiteral ;
                  oslc:propertyDefinition :argument ;
                  oslc:name "argument"^^xsd:string ;
                  oslc:occurs oslc:Exactly-one ;
                  oslc:valueType oslc:AnyResource
                ] .
**There should be a :FluentPlusShape as well**

:QuantifierShape
  a oslc:ResourceShape ;
  oslc:describes :Quantifier ;
  dcterms:description "Quantifiers (forall,exists) in precondition."^^rdf:XMLLiteral ;
:ForAllShape
  a oslc:ResourceShape ;
  oslc:describes :ForAll ;
  dcterms:description "Universal quantifier (forall) in effect."^^rdf:XMLLiteral ;

**Do we need the ForAllShape ? **

**Comment by likelion** [comment]
@swarupmohalik:
1. According to https://www.jair.org/media/1129/live-1129-2132-jair.pdf (page 119), it is possible to have multiple entries in `init` (I guess with implicit conjunctive semantics):
```
<init> ::= (:init <init-el>∗)
<init-el> ::= <literal(name)>
<init-el> ::= (= <f-head> <number>)
```
2. I didn't put a `FluentPlusShape` because it doesn't restrict anything. According to OSLC shapes it is not possible to set `oslc:occurs` to exactly `2` and it is not possible to restrict arguments with `oslc:range` to  anything reasonable because inferencing over `rdfs:subClassOf` is not allowed. I only left resource shapes that can restrict at least something, e.g. for `FluentMinusShape` they restrict occurrences to exactly 1 in case of unary `-` operator.

3. There is `ForAllShape` for the case when it is used in effect, otherwise see my point 2 (such a shape will not restrict anything).

I know, shapes are not perfect to perform complete validation of PDDL in RDF documents, but this is what they can do. If we want to impose stricter constraints, we need to use some other technology (e.g. SHACL), but there is no implementation of it for Prolog, and I don't have time at the moment to implement it. In any case, resource shapes are optional and everything will work without them, plus there are additional constraints imposed by the code to perform more validation during transformation operations.

**Comment by likelion** [comment]
Planner service base URI:  `http://<host>/planner/planCreationFactory`

**Generate PDDL text:** do `POST` with `Accept: text/x-pddl` header and body containing either `pddl:Domain` or `pddl:Problem` or `pddl:Plan` resources or any combination of them.

**Generate plan:** do `POST` with `Accept: application/turtle` or `application/rdf+xml` header and body containing both `pddl:Domain` and `pddl:Problem` resources. The returned resource is of type `pddl:Plan` containing steps as resources of type `pddl:Step` ordered using `sh:order` property.

For plan generation Metric-FF planner binary `ff` should be where `run.pl` or in `$PATH`.

**Comment by swarupmohalik** [comment]
In scott-eu/planner_reasoner/rdf/base/pddl_example.ttl

 [ a pddl:Increase ;
                                pddl:parameter [ a :moved ;
                                                 :moved-m :move-b
                                               ] ;
                                pddl:argument 1
                              ] ,
 [ a pddl:Increase ;
                                pddl:parameter [ a :total-moved ] ;
                                pddl:argument 1
                              ] ,
**"1" should be an expression (f-exp)**

image

image

**Comment by swarupmohalik** [comment]
I think the conditional effects (When) are not fully supported, am I right? 
It is alright for the time being since the warehouse domain does not use conditional effects.

**Comment by swarupmohalik** [comment]
<init> ::= (:init <init-el>∗)

There is only one :init block in a problem. 
Multiple <init-el>s can be there in the block.

**Comment by swarupmohalik** [comment]
[ a oslc:Property ;
                  dcterms:description "Problem init."^^rdf:XMLLiteral ;
                  oslc:propertyDefinition :init ;
                  oslc:name "init"^^xsd:string ;
                  oslc:occurs oslc:Zero-or-many ;
                  oslc:valueType oslc:AnyResource
                ] ,

**I was referring to this earlier. The shape allows multiple init blocks.
the occurs should be exactly-one**

**Comment by likelion** [comment]
@swarupmohalik

The `f-exp` is an abstract syntax node, so it is not specified in the input RDF file, but it is properly handled by the PDDL generator (see pddl_generator.pl, which has `f_exp` rule). One of the possible contents of `f_exp` is a number, the type can be inferred from the xsd literal type. So, the example file is correct.

The conditional effect using `when` construct is supported by the PDDL generator completely, but actual operational support is up to a planning engine that we plug in.

Regarding `init`: your interpretation is correct.

**Comment by likelion** [comment]
@swarupmohalik

The resource shape of `init` property allows multiple values of the property, not multiple `init` specifications. The PDDL generator always generates single `init` with all property values combined in it.

In general, resource shapes have nothing to do with syntax of PDDL files, they specify allowed properties of RDF files.

**Comment by kicacacko** [comment]
Done review on the PDDL ontology...

Considering Parameters. I like that there are no variables, and it is solved everything with parameters. Note that the order of parameters in an action is not important, however as you noted the order in a predicate or a function is important. Since there is only one ParameterShape i am ok with having sh:order (or any other mechanism to model the position) property there, even when not used. 

On a side note, currently the plan is described as a set of steps using StepShape. Swarup and me discussed that we should include a plan validator tool in the planner adapter (probably VAL https://nms.kcl.ac.uk/planning/software/val.html) that will take the plan and translate it in a format of add/delete predicates, so that during runtime the Estimator knows whether the current state is equvalent to the planned state.

**Review by kicacacko — APPROVED** [review]

**Pull request merged**